### DISCUSSION TITLE
Is it / could it be possible for Flex to handle labels?

### DISCUSSION THREAD

**Discussion by crumblingstatue** [fltk-rs/fltk-rs#938] [Q&A] [answered]
Flex is a major improvement over manually specifying widget sizes.
There is, however one thing that still bothers me. It doesn't take into account labels.
Take this for example
```rust
let row = Flex::default().with_type(FlexType::Row);
Input::default().with_label("Input 1");
Input::default().with_label("Input 2");
row.end();
```

It ends up looking like this:
![image](https://user-images.githubusercontent.com/1521976/137207918-06e20f51-05f2-4edd-8b00-c47c62f862b1.png)

Would it be possible to have Flex take into account labels? And if not, what's the most ergonomic option to deal with them?

**Comment by crumblingstatue** [comment]
One solution is to use Frames like this
```Rust
let row = Flex::default().with_type(FlexType::Row);
Frame::default().with_label("Input 1");
Input::default();
Frame::default().with_label("Input 2");
Input::default();
row.end();
```
However, they don't "stick" to the inputs and they take up a lot of space.
![image](https://user-images.githubusercontent.com/1521976/137208405-a1d7f79c-deaf-469b-9eec-892d57254ea2.png)

**Comment by MoAlyousef** [comment, marked as answer]
Currently unfortunately no. I think I'll try to work on it by basically checking the label size (using measure_label, and checking the label's alignment, then taking that into consideration with the Flex calculation).
Currently, you can either use a Flex with labels and label-less inputs, while specifying the alignment of the labels:
```rust
	let row = Flex::default().with_type(FlexType::Row);
	Frame::default().with_label("Input 1").with_align(Align::Right | Align::Inside);
	Input::default();
	Frame::default().with_label("Input 2").with_align(Align::Right | Align::Inside);
	Input::default();
	row.end();
```
![image](https://user-images.githubusercontent.com/37966791/137214081-b7e450da-3a9b-47ea-a698-ebd0aed1c52d.png)

Or you can use a Pack, and specify the width of the inputs and a spacing:
```rust
	let mut row = Pack::default().with_type(PackType::Horizontal);
	row.set_spacing(70);
	Input::default().with_size(100, 0).with_label("Input 1");
	Input::default().with_size(100, 0).with_label("Input 2");
	row.end();
```
![image](https://user-images.githubusercontent.com/37966791/137214648-0b083ab2-02e5-4d82-a15d-64ea8eaf7a72.png)

Packs also have an auto_layout method, which takes into account the spacing, unlike the Flex widget:
```rust
	let mut row = Pack::default().with_type(PackType::Horizontal);
	row.set_spacing(70);
	Input::default().with_label("Input 1");
	Input::default().with_label("Input 2");
	row.end();
	row.auto_layout();
```
So no need to specify the sizes, just the spacing and call auto_layout().

**Comment by MoAlyousef** [comment]
Another different option is to change the alignment of the input itself:
```rust
	let mut row = Flex::default().with_type(FlexType::Row);
	Input::default().with_label("Input 1").with_align(Align::TopLeft);
	Input::default().with_label("Input 2").with_align(Align::TopLeft);
	row.end();
```
![image](https://user-images.githubusercontent.com/37966791/137219312-6462a037-10e8-4fde-8f43-69213b0c9656.png)

**Reply by crumblingstatue** [reply]
Thank you for the detailed answers!
